runtime/protocol/http-client/jvm/test/aws/smithy/kotlin/runtime/http/interceptors/requestcompression/GzipCompressionTest.kt

```diff
@@ -0,0 +1,180 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package aws.smithy.kotlin.runtime.http.interceptors.requestcompression
+
+import aws.smithy.kotlin.runtime.hashing.crc32
+import aws.smithy.kotlin.runtime.http.interceptors.decompressGzipBytes
+import aws.smithy.kotlin.runtime.io.*
+import kotlinx.coroutines.runBlocking
+import kotlinx.coroutines.test.runTest
+import org.junit.jupiter.params.ParameterizedTest
+import org.junit.jupiter.params.provider.ValueSource
+import kotlin.test.assertContentEquals
+import kotlin.test.assertEquals
+
+class GzipCompressionTest {
```

Comment:
**suggestion**: You have a good start here and as Matas said I like that you are trying to test different read methods. I think this test could be a lot cleaner if you refactor it to something like https://github.com/awslabs/smithy-kotlin/blob/ed7cad2d4f3f5d2d244efee9315b28a5ce63c148/runtime/runtime-core/common/test/aws/smithy/kotlin/runtime/content/ByteStreamFlowTest.kt#L14-L18

Define an abstract test that has all the different content you want to compress and different methods of reading said content and instantiate it with different inputs (`SdkSource`, `SdkByteReadChannel`, `ByteArray`, etc). You'll have to create some utility types to do this but it would look something like:


```kotlin

fun interface GzipTestReader {
    suspend fun read(sink: SdkBuffer, limit: Long): Long
}

fun interface GzipTestReaderFactory {
    fun newReader(payload: ByteArray): GzipTestReader
    
    companion object {
        val SDK_SOURCE = GzipTestReaderFactory { payload ->
            val source = GzipSdkSource(payload.source(), payload.size.toLong())
            GzipTestReader { sink, limit ->  
                source.read(sink, limit)
            }
        }
        
        val BYTE_CHANNEL = GzipTestReaderFactory { payload ->
            val ch = SdkByteReadChannel(payload)
            val source = GzipByteReadChannel(ch)
            GzipTestReader { sink, limit ->
                source.read(sink, limit)
            }
        }
    }
}

class GzipSourceCompressionTest: GzipCompressionTest(GzipTestReaderFactory.SDK_SOURCE)
class GzipChannelCompressionTest: GzipCompressionTest(GzipTestReaderFactory.BYTE_CHANNEL)

abstract class GzipCompressionTest(
    val factory: GzipTestReaderFactory
) {
    @Test
    fun testReadLargeBody() {
        // use factory to create new instances and test read against it

    }

    @Test
    fun testReadChunked() {

    }

    @Test
    fun testReadOneByteAtTime() {

    }

    // ...


}

```

The advantage of doing this is being able to debug specific tests better as well as getting the benefits of parameterization and test scenario re-use across different readers.